optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/incremental/IncrementalScoreDirector.java

```diff
@@ -42,6 +42,12 @@ public IncrementalScoreDirector(IncrementalScoreDirectorFactory scoreDirectorFac
                                     boolean constraintMatchEnabledPreference,
                                     IncrementalScoreCalculator<Solution_> incrementalScoreCalculator) {
         super(scoreDirectorFactory, constraintMatchEnabledPreference);
+        if (constraintMatchEnabledPreference
+                && !(incrementalScoreCalculator instanceof ConstraintMatchAwareIncrementalScoreCalculator)) {
+            throw new IllegalStateException("The score calculator " + incrementalScoreCalculator + " must implement "
+                    + ConstraintMatchAwareIncrementalScoreCalculator.class.getSimpleName()
+                    + " when constraintMatchEnabledPreference is true.");
+        }
```

Comment:
I fear we won't be able to add this fail fast (although I always like any proposal to add more fail fasts :)). The other fail fasts added in the PR are fine.